src/mlpack/core/data/string_encoding_policies/tf_idf_encoding_policy.hpp

```diff
@@ -0,0 +1,196 @@
+ * from one. The tokens are labeled in the order of their occurrence 
+ * in the input dataset.
+ */
+class TfIdfEncodingPolicy
+{
+ public:
+
+  TfIdfEncodingPolicy(bool binary = false, bool smooth_idf = true,
+                      bool sublinear_tf = false, bool raw_count = true,
+                      bool term_frequency = false) : binary (binary),
+                      smooth_idf (smooth_idf), sublinear_tf (sublinear_tf),
+                      raw_count (raw_count), term_frequency (term_frequency) {}
+  /**
+  * The function initializes the output matrix.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param datasetSize The number of strings in the input dataset.
+  * @param maxNumTokens The maximum number of tokens in the strings of the 
+                        input dataset.
+  * @param dictionarySize The size of the dictionary (not used).
+  */
+  template<typename MatType>
+  static void InitMatrix(MatType& output,
+                         size_t datasetSize,
+                         size_t /*maxNumTokens*/,
+                         size_t dictionarySize)
+  {
+    output.zeros(datasetSize, dictionarySize);
+  }
+
+  /**
+  * The function initializes the output matrix.
+  * Overloaded function to store result in vector<vector<size_t>>
+  * 
+  * @param output Output matrix to store the encoded results.
+  * @param datasetSize The number of strings in the input dataset.
+  * @param maxNumTokens The maximum number of tokens in the strings of the 
+                        input dataset.
+  * @param dictionarySize The size of the dictionary (not used).
+  */
+  template<typename OutputType>
+  static void InitMatrix(std::vector<std::vector<OutputType> >& output,
+                         size_t datasetSize,
+                         size_t /*maxNumTokens*/,
+                         size_t dictionarySize)
+  {
+    output.resize(datasetSize, std::vector<OutputType> (dictionarySize, 0));
+  }
+
+  /** 
+  * The function performs the TfIdf encoding algorithm i.e. it writes
+  * the encoded token to the ouput.
+  *
+  * @param output Output matrix to store the encoded results (sp_mat or mat).
+  * @param value The encoded token.
+  * @param row The row number at which the encoding is performed.
+  * @param col The row token number at which the encoding is performed.
+  */
+  template<typename MatType>
+  void Encode(MatType& output, size_t value, size_t row, size_t /*col*/)
+  {
+    // Important since Mapping starts from 1 whereas allowed column value is 0.
+    double idf, tf;
+    if(smooth_idf)
+      idf = std::log((output.n_rows + 1) / (1 + idfdict[value-1])) + 1;
+    else
+      idf = std::log(output.n_rows / idfdict[value - 1]) + 1;
+    if (sublinear_tf)
+      tf = tokenCount[row][value - 1] / row_size[row];
+    else if (term_frequency)
+      tf = std::log(tokenCount[row][value - 1]) + 1;
+    else if (binary)
+      tf = tokenCount[row][value - 1] > 0 ? 1 : 0;
+    else
```

Comment:
If one excludes the other we should use `enum class` instead of `bool`.

Probably we should use a template parameter instead since the comparison is performed inside the loop.